erpnext/stock/doctype/warehouse/warehouse.json

```diff
@@ -70,9 +71,9 @@
    "oldfieldname": "company",
    "oldfieldtype": "Link",
    "options": "Company",
-   "remember_last_selected_value": 1,
    "reqd": 1,
-   "search_index": 1
+   "search_index": 1,
+   "set_only_once": 1
```

Comment:
I believe `set_only_once` just makes sure the value of this field never changes after the first insert. it does not impact the field visibility.

What you should additionally do is:
-  make this field mandatory as it is an important field
- make **Read Only Depends on** as `eval: !doc.__islocal`. This means while the document is new, unsaved and un-inserted for the first time, the field should be editable. After the first save it should be read only